timemachine/cpp/src/wrap_kernels.cpp

```diff
@@ -308,6 +309,107 @@ void declare_context(py::module &m) {
+                    throw std::runtime_error("k must be at least one");
+                }
+                // TBD determine a more precise threshold, currently 10x what has been tested
+                const double max_k = 1000000.0;
+                if (k > max_k) {
+                    throw std::runtime_error("k must be less than than " + std::to_string(max_k));
+                }
+
+                const int N = ctxt.num_atoms();
+                const int x_interval = (store_x_interval <= 0) ? n_steps : store_x_interval;
+
+                std::vector<int> vec_selection_mask(selection_mask.size());
+                std::memcpy(vec_selection_mask.data(), selection_mask.data(), vec_selection_mask.size() * sizeof(int));
+                verify_atom_idxs(N, vec_selection_mask);
+
+                std::array<std::vector<double>, 2> result =
+                    ctxt.multiple_steps_local_selection(n_steps, reference_idx, vec_selection_mask, burn_in, x_interval, radius, k);
+                const int D = 3;
+                const int F = result[0].size() / (N * D);
+                py::array_t<double, py::array::c_style> out_x_buffer({F, N, D});
+                std::memcpy(out_x_buffer.mutable_data(), result[0].data(), result[0].size() * sizeof(double));
+
+                py::array_t<double, py::array::c_style> box_buffer({F, D, D});
+                std::memcpy(box_buffer.mutable_data(), result[1].data(), result[1].size() * sizeof(double));
+                return py::make_tuple(out_x_buffer, box_buffer);
+            },
+            py::arg("n_steps"),
+            py::arg("reference_idx"),
+            py::arg("selection_mask"),
+            py::arg("burn_in") = 500, // This is arbitrarily selected as a default, TODO make informed choice
+            py::arg("store_x_interval") = 0,
+            py::arg("radius") = 1.2,
+            py::arg("k") = 10000.0,
+            R"pbdoc(
+        Take multiple steps using a selection of free particles restrainted to a reference particle. Useful for avoiding the bias
+        introduced by switching on and off the restraint on different particles as is done with multiple_steps_local.
+
+        Running a barostat and local MD at the same time are not currently supported. If a barostat is
+        assigned to the context, the barostat won't run.
+
+        Note: Running this multiple times with small number of steps (< 100) may result in a vacuum around the local idxs due to
+        discretization error caused by switching on the restraint after a particle has moved beyond the radius.
+
+        F = iterations / store_x_interval
+
+        The first call to `multiple_steps_local_selection` takes longer than subsequent calls, if initialize_local_md has not been called previously,
+        initializes potentials needed for local md.
+
+        Parameters
+        ----------
+        n_steps: int
+            Number of steps to run.
+
+        selection_mask: np.array of int32
```

Comment:
nit [ future / naming ] -- `selection_mask` -> `selection_idxs`.

(To standardize on `selection_mask` = length-N boolean mask, `selection_mask[i] = True` if `i` is selected, vs. `selection_idxs` = variable length int list, `i in selection_idxs` if `i` is selected.)

Reply:
+1 - I got confused by this too. I like the convention that `masks` are boolean arrays, and `idxs` are list of ints.

Reply:
(note that what's more confusing, is that we often times avoid using `std::vector<bool>` to denote boolean flags of size `N` particles, i.e. we still prefer `std::vector<int>`, due to `sizeof(bool)` being a bit janky)

Reply:
In either case, I think documenting the size of `selection_mask` would be useful, i.e. either being of `size K` or something else

Reply:
Done in https://github.com/proteneer/timemachine/pull/1005/commits/7eee6dec2226057d4513b3770313fa89d2e6d521

Though didn't document size, will expand that

Reply:
Added in https://github.com/proteneer/timemachine/pull/1005/commits/1b098ad9cacf81bf0b7c5d0ad4bf664a92901b59